lib/srv/desktop/rdp/rdpclient/src/rdpdr/filesystem.rs

```diff
@@ -268,6 +273,85 @@ impl FilesystemBackend {
+                ));
+            }
+
+            if rdp_req.initial_query == 0 {
+                // This isn't the initial query, ergo we already have this dir's contents filled in.
+                // Just send the next item.
+                return self.send_next_drive_query_dir_response(&rdp_req);
+            }
+
+            // On the initial query, we need to get the list of files in this directory from
+            // the client by sending a TDP SharedDirectoryListRequest.
+            // https://github.com/FreeRDP/FreeRDP/blob/511444a65e7aa2f537c5e531fa68157a50c1bd4d/channels/drive/client/drive_file.c#L775
+            let path = dir.path.clone();
+
+            // Ask the client for the list of files in this directory.
+            self.send_tdp_sd_list_request(tdp::SharedDirectoryListRequest {
+                completion_id: rdp_req.device_io_request.completion_id,
+                directory_id: rdp_req.device_io_request.device_id,
+                path,
+            })?;
+
+            // When we get the response for that list of files...
+            self.pending_sd_list_resp_handlers.insert(
+                rdp_req.device_io_request.completion_id,
+                SharedDirectoryListResponseHandler::new(
+                    move |cli: &mut Self,
+                          tdp_resp: tdp::SharedDirectoryListResponse|
+                          -> PduResult<()> {
+                        if tdp_resp.err_code != TdpErrCode::Nil {
+                            // For now any error will kill the session.
+                            // In the future, we might want to make this send back
+                            // an NTSTATUS::STATUS_UNSUCCESSFUL instead.
+                            return Err(custom_err!(
+                                "FilesystemBackend::handle_query_directory_req",
+                                FilesystemBackendError(format!(
+                                    "SharedDirectoryListRequest failed with err_code = {:?}",
+                                    tdp_resp.err_code
+                                ))
+                            ));
+                        }
+
+                        // If SharedDirectoryListRequest succeeded, move the
+                        // list of FileSystemObjects that correspond to this directory's
+                        // contents to its entry in the file cache.
+                        if let Some(dir) = cli.file_cache.get_mut(file_id) {
+                            dir.contents = tdp_resp.fso_list;
+                            // And send back the "." directory over RDP
+                            return cli.send_next_drive_query_dir_response(&rdp_req);
+                        }
+
+                        cli.send_drive_query_dir_response(
+                            rdp_req.device_io_request,
+                            NtStatus::UNSUCCESSFUL,
+                            None,
+                        )
```

Comment:
Can we extract this to separate method? It's really hard to read when it's so deeply nested

Reply:
+1, and consider handling the error case (file not found in cache) up front.

Reply:
b4518cd604b3c2b969ee16488e1d72118c4c0203